File: publicq-main/src/PublicQ.Infrastructure/Options/IpRateLimitOptions.cs
namespace PublicQ.Infrastructure.Options;

/// <summary>
/// Model for IP rate limiting options.
/// It has to be in sync with AspNetCoreRateLimit nuget package.
/// </summary>
public class IpRateLimitOptions
{
    /// <summary>
    /// Array of <see cref="GeneralRules"/>
    /// </summary>
    public List<GeneralRule> GeneralRules { get; set; } = [];
    
    /// <summary>
    /// Array of IP addresses (v4 and v6) that are exempt from rate limiting.
    /// </summary>
    public List<string> IpWhitelist { get; set; } = [];
    
    /// <summary>
    /// Real IP header to use when behind a proxy.
    /// </summary>
    public string RealIpHeader { get; set; } = "X-Real-IP";
}

/// <summary>
/// General rule for rate limiting.
/// </summary>
public class GeneralRule
{
    /// <summary>
    /// Endpoint to which the rule applies. Use "*" to apply to all endpoints.
    /// </summary>
    public string Endpoint { get; set; } = default!;
    
    /// <summary>
    /// API calls limit for the specified period.
    /// </summary>
    public int Limit { get; set; }
    
    /// <summary>
    /// Period for the rate limit (e.g., "5s" for 5 seconds, "1m" for 1 minute).
    /// </summary>
    public string Period { get; set; } = default!;
}

File: publicq-main/client/src/models/ip-rate-limit-options.ts
/**
 * IP rate limiting configuration options.
 */
export interface IpRateLimitOptions {
  /**
   * Enable endpoint rate limiting.
   */
  enableEndpointRateLimiting: boolean;

  /**
   * Stack blocked requests.
   */
  stackBlockedRequests: boolean;

  /**
   * HTTP status code to return when rate limit is exceeded.
   */
  httpStatusCode: number;

  /**
   * Header used to determine the real client IP when behind a proxy.
   */
  realIpHeader: string;

  /**
   * A list of IP addresses (v4 and v6) that are exempt from rate limiting.
   */
  ipWhitelist: string[];

  /**
   * A list of endpoint patterns that are exempt from rate limiting.
   */
  endpointWhitelist: string[];

  /**
   * General rate limiting rules applied to all endpoints.
   */
  generalRules: RateLimitRule[];

  /**
   * Rate limiting rules for specific endpoints.
   */
  rules: RateLimitRule[];
}

/**
 * Rate limiting rule configuration.
 */
export interface RateLimitRule {
  /**
   * Endpoint pattern (e.g., "*", "/api/*", "/api/users").
   */
  endpoint: string;

  /**
   * Time period for the rate limit (e.g., "1s", "1m", "1h").
   */
  period: string;

  /**
   * Maximum number of requests allowed in the specified period.
   */
  limit: number;
}

/**
 * Default IP rate limiting options.
 */
export const defaultIpRateLimitOptions: IpRateLimitOptions = {
  enableEndpointRateLimiting: false,
  stackBlockedRequests: false,
  httpStatusCode: 429,
  realIpHeader: "X-Real-IP",
  ipWhitelist: [],
  endpointWhitelist: [],
  generalRules: [],
  rules: []
};